### ISSUE TITLE
holdApplicationUntilProxyStarts breaks rewriteAppPropers

### ISSUE THREAD

**Issue by saltenhub** [istio/istio#26873]
**Bug description**
Recently upgraded to v1.7.0 (upgraded over multiple upgrades from ~1.4.2). No major custom configuration besides `values.global.proxy.holdApplicationUntilProxyStarts=true`, but when I try to enable mutual TLS on a namespace with a `PeerAuthentication`, the healthchecks start to fail (even with the `samples/health-check/liveness-http.yaml` example). `rewriteAppHTTPProbers` is enabled cluster-wide and changing it in the pod annotations also does not yield a change.

I get these logs in the istio-proxy container before the pod is terminated an restarted leading to a CrashLoopbackOff.

```
2020-08-27T21:40:03.205178Z	error	Request to probe app failed: Get "http://localhost:8001/foo": dial tcp 127.0.0.1:8001: connect: connection refused, original URL path = /app-health/liveness-http/livez
app URL path = /foo
```

[ ] Docs
[ ] Installation
[x] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Expected behavior**
Istio mutal TLS working with liveness and readiness probes.

**Steps to reproduce the bug**
Upgrade to version 1.7.0 and try out to enable mTLS in a namespace using `PeerAuthentication`.

**Version (include the output of `istioctl version --remote` and `kubectl version` and `helm version` if you used Helm)**
1.7.0

**How was Istio installed?**
istioctl install / upgrade

I'm really not sure if this is because of the upgrading of one version to the next, but I would really be thankful for some help. I am happy to provide additional data on request.

**Update 08/28/20 10:55 UTC**:
Forgot to mention `values.global.proxy.holdApplicationUntilProxyStarts=true` was set during installation. Added it in the description

**Comment by howardjohn** [comment]
Does your application listen on <POD_IP> instead of 0.0.0.0 or localhost?

Can you do `curl localhost:8001/foo` from the app container and/or the sidecar container?

**Comment by saltenhub** [comment]
I can curl from the sidecar container (the liveness-http sample does not seem to provide curl):
```
keti -n test liveness-http-64c6856745-lxjvg -c istio-proxy -- curl http://localhost:8001/foo
-> Listening on 8001: foo % 
```

**Comment by dntosas** [comment]
same problem in here when enabling `values.global.proxy.holdApplicationUntilProxyStarts=true`

**Comment by saltenhub** [comment]
@dntosas You're absolutely right, I have enabled that as well but forgot to mention it. Will edit original issue to reflect that.

**Comment by dntosas** [comment]
i think we got something

when enabling `proxy.holdApplicationUntilProxyStarts=true` looking at `istio-proxy` container template:
```
    Liveness:   http-get http://:15020/app-health/papyrus/livez delay=0s timeout=10s period=10s #success=1 #failure=6
    Readiness:  http-get http://:15020/app-health/papyrus/readyz delay=0s timeout=10s period=10s #success=1 #failure=6
```

when disabling `proxy.holdApplicationUntilProxyStarts=false` looking at `istio-proxy` container template:
```
    Readiness:  http-get http://:15021/healthz/ready delay=1s timeout=1s period=2s #success=1 #failure=30
```

the difference is that no **Liveness** probe is present when this feature is disabled

cc @howardjohn @luksa

**Comment by dntosas** [comment]
Maybe we got something

When **DISABLED** `global.proxy.holdApplicationUntilProxyStarts` looking at container template using `kubectl describe pod`:
```
istio-proxy:
  Readiness:  http-get http://:15021/healthz/ready delay=1s timeout=1s period=2s #success=1 #failure=30
```

When **ENABLED** `global.proxy.holdApplicationUntilProxyStarts` looking at container template using `kubectl describe pod`:
```
istio-proxy:
    Liveness:   http-get http://:15020/app-health/MYAPP/livez delay=0s timeout=10s period=10s #success=1 #failure=6
    Readiness:  http-get http://:15020/app-health/MYAPP/readyz delay=0s timeout=10s period=10s #success=1 #failure=6
```

The are differences seem to be present on how **Liveness/Readiness** is defined when feature is enabled ^^

cc @howardjohn @luksa

**Comment by howardjohn** [comment]
I can reproduce this

@luksa I think its applying the rewrite app probe to `container[0]` or something - which is now the sidecar

**Comment by dntosas** [comment]
> I can reproduce this
> 
> @luksa I think its applying the rewrite app probe to `container[0]` or something - which is now the sidecar

Ref --> https://github.com/istio/istio/blob/1cc05bcf798644262547032d6714fab2603121e3/pkg/kube/inject/app_probe.go#L142-L144

**Comment by tariq1890** [comment]
@dntosas Does this work when ` holdApplicationUntilProxyStarts` is set to false?

**Comment by tariq1890** [comment]
I think I figured out the issue for this.

Sending a fix shortly

**Comment by tariq1890** [comment]
Just wanted to send an update. It looks like the fix is far more complicated than I thought. Fixing this would require more thought since we face the risk of setting of regressions.

**Comment by tariq1890** [comment]
This can be closed

**Comment by danquack** [comment]
@tariq1890 is this closed because that PR resolves it? Should we avoid implementing this until the next 1.7.1?

**Comment by tariq1890** [comment]
@danquack Yes. I had tested the fix for it in release-1.7 alpha and it works as expected. You can try this feature out in 1.7.1

**Comment by dntosas** [comment]
FYI, tested the fix delivered with `1.7.1` and works as expected :tada:
